receiver/prometheusreceiver/metrics_receiver.go

```diff
@@ -61,16 +73,129 @@ func (r *pReceiver) Start(_ context.Context, host component.Host) error {
 
 	logger := internal.NewZapToGokitLogAdapter(r.settings.Logger)
 
-	discoveryManager := discovery.NewManager(discoveryCtx, logger)
+	// add scrape configs defined by the collector configs
+	baseCfg := r.cfg.PrometheusConfig
+
+	err := r.initPrometheusComponents(discoveryCtx, host, logger)
+	if err != nil {
+		r.settings.Logger.Error("Failed to initPrometheusComponents Prometheus components", zap.Error(err))
+		return err
+	}
+
+	err = r.applyCfg(baseCfg)
+	if err != nil {
+		r.settings.Logger.Error("Failed to apply new scrape configuration", zap.Error(err))
+		return err
+	}
+
+	allocConf := r.cfg.TargetAllocator
+	if allocConf != nil {
+		go func() {
+			// immediately sync jobs and not wait for the first tick
+			savedHash, _ := r.syncTargetAllocator(uint64(0), allocConf, baseCfg)
+			r.targetAllocatorIntervalTicker = time.NewTicker(allocConf.Interval)
+			for {
+				<-r.targetAllocatorIntervalTicker.C
+				hash, err := r.syncTargetAllocator(savedHash, allocConf, baseCfg)
+				if err != nil {
+					r.settings.Logger.Error(err.Error())
+					continue
+				}
+				savedHash = hash
+			}
+		}()
+	}
+
+	return nil
+}
+
+// syncTargetAllocator request jobs from TargetAllocator and update underlying receiver, if the response does not match the provided compareHash.
+// baseDiscoveryCfg can be used to provide additional ScrapeConfigs which will be added to the retrieved jobs.
+func (r *pReceiver) syncTargetAllocator(compareHash uint64, allocConf *TargetAllocator, baseCfg *config.Config) (uint64, error) {
+	r.settings.Logger.Debug("Syncing target allocator jobs")
+	jobObject, err := getJobResponse(allocConf.Endpoint)
+	if err != nil {
+		r.settings.Logger.Error("Failed to retrieve job list", zap.Error(err))
+		return 0, err
+	}
+
+	hash, err := hashstructure.Hash(jobObject, hashstructure.FormatV2, nil)
+	if err != nil {
+		r.settings.Logger.Error("Failed to hash job list", zap.Error(err))
+		return 0, err
+	}
+	if hash == compareHash {
+		// no update needed
+		return hash, nil
+	}
+
+	cfg := baseCfg
+
+	for _, linkJSON := range *jobObject {
+		httpSD := *allocConf.HTTPSDConfig
+		httpSD.URL = fmt.Sprintf("%s%s?collector_id=%s", allocConf.Endpoint, linkJSON.Link, allocConf.CollectorID)
+
+		scrapeCfg := &config.ScrapeConfig{
+			JobName: linkJSON.Link,
+			ServiceDiscoveryConfigs: discovery.Configs{
+				&httpSD,
+			},
+		}
+
+		cfg.ScrapeConfigs = append(cfg.ScrapeConfigs, scrapeCfg)
```

Comment:
Can this conflict with the base configuration provided?  How should we handle those conflicts?

I almost wonder if this should be a separate receiver that wraps the existing prometheus receiver.

Reply:
where do the configurations get merged? Is that just prometheus logic for replacing duplicate jobs? In which case is it first in wins?

Reply:
I think that even if there was no base configuration that updates from the target allocator could cause problems over time.  If a job is added then removed then restored it could appear twice.  Some manner of de-duplication is probably necessary.

Reply:
I save the baseConfig and use it as basis at each time a sync is necessary. https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12586/files#diff-8172ccc732b059ffa8d9d8858643701e32c90c3d3ab8b6697fa74ca1c8c09fb5R133

Reply:
Doesn't that just take a shallow copy?  `cfg.ScrapeConfigs` is a slice, which is a reference type and two structs holding a slice with the same underlying type can be entangled, no?  I think we need a test that ensures there is no entanglement here.

Reply:
I haved a test which covers changes in the slice.

Reply:
Can you open an issue to ensure we deduplicate targetallocator-provided scrape configs and base config scrape configs?  The failure mode is harsh here:  If the base config conflicts with the TA-provided config, the collector will not receive any more config updates